hugr-model/src/v0/mod.rs

```diff
@@ -0,0 +1,621 @@
+    /// The parameters of the node.
+    pub params: &'a [TermId],
+    /// The regions of the node.
+    pub regions: &'a [RegionId],
+    /// The meta information attached to the node.
+    pub meta: &'a [MetaItem<'a>],
+    /// The type of the node.
+    pub r#type: TermId,
+}
+
+/// Operations that nodes can perform.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum Operation<'a> {
+    /// Data flow graphs.
+    Dfg,
+    /// Control flow graphs.
+    Cfg,
+    /// Basic blocks.
+    Block,
+    /// Function definitions.
+    DefineFunc {
+        /// The declaration of the function to be defined.
+        decl: &'a FuncDecl<'a>,
+    },
+    /// Function declarations.
+    DeclareFunc {
+        /// The function to be declared.
+        decl: &'a FuncDecl<'a>,
+    },
+    /// Function calls.
+    CallFunc {
+        /// The function to be called.
+        func: TermId,
+    },
+    /// Function constants.
+    LoadFunc {
+        /// The function to be loaded.
+        func: TermId,
+    },
+    /// Custom operation.
+    ///
+    /// The implicit parameters of the operation are left out.
+    Custom {
+        /// The name of the custom operation.
+        name: GlobalRef<'a>,
+    },
+    /// Custom operation.
+    ///
+    /// The implicit parameters of the operation are included.
+    CustomFull {
+        /// The name of the custom operation.
+        name: GlobalRef<'a>,
+    },
+    /// Alias definitions.
+    DefineAlias {
+        /// The declaration of the alias to be defined.
+        decl: &'a AliasDecl<'a>,
+        /// The value of the alias.
+        value: TermId,
+    },
+
+    /// Alias declarations.
+    DeclareAlias {
+        /// The alias to be declared.
+        decl: &'a AliasDecl<'a>,
+    },
+
+    /// Tail controlled loop.
+    /// Nodes with this operation contain a dataflow graph that is executed in a loop.
+    /// The loop body is executed at least once, producing a result that indicates whether
+    /// to continue the loop or return the result.
+    ///
+    /// # Port Types
+    ///
+    /// - **Inputs**: `inputs` + `rest`
+    /// - **Outputs**: `outputs` + `rest`
+    /// - **Sources**: `inputs` + `rest`
+    /// - **Targets**: `(adt [inputs outputs])` + `rest`
+    TailLoop,
+
+    /// Conditional operation.
+    ///
+    /// # Port types
+    ///
+    /// - **Inputs**: `[(adt inputs)]` + `context`
+    /// - **Outputs**: `outputs`
+    Conditional,
+
+    /// Create an ADT value from a sequence of inputs.
+    Tag {
+        /// The tag of the ADT value.
+        tag: u16,
+    },
+}
+
+/// A region in the hugr.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Region<'a> {
+    /// The kind of the region. See [`RegionKind`] for details.
+    pub kind: RegionKind,
+    /// The source ports of the region.
+    pub sources: &'a [Port<'a>],
+    /// The target ports of the region.
+    pub targets: &'a [Port<'a>],
+    /// The nodes in the region. The order of the nodes is not significant.
```

Comment:
not clear what the ports of a region are. In dataflow, are they the ports belonging to Input/Output nodes?

Reply:
is this why Input/Output aren't Operation variants?